### PR TITLE
Pipe: validate the value of extractor.pattern

### PR DESCRIPTION
**Pull request by xuanronaldo**
As title. Now only strings that meet the IoTDB time series path rules can become parameter values for extractor.pattern.
It will throw an exception like this, if the pattern is an illegal path:
image

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java [hunk 1]
```diff
@@ -84,6 +87,11 @@ public IoTDBDataRegionExtractor() {
 
   @Override
   public void validate(PipeParameterValidator validator) throws Exception {
+    // Check whether the pattern is legal
+    PathUtils.isLegalPath(
+        validator
+            .getParameters()
+            .getStringOrDefault(EXTRACTOR_PATTERN_KEY, EXTRACTOR_PATTERN_DEFAULT_VALUE));
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java:94]:
```suggestion
    // Check whether the pattern is legal
    PathUtils.isLegalPath(
        validator
            .getParameters()
            .getStringOrDefault(EXTRACTOR_PATTERN_KEY, EXTRACTOR_PATTERN_DEFAULT_VALUE));

```

**Review by SteveYurongSu — APPROVED** [review]

File: iotdb-core/datanode/src/test/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractorTest.java [hunk 2]
```diff
@@ -53,4 +54,21 @@ public void testIoTDBDataRegionExtractor() {
...
+      extractor.validate(
+          new PipeParameterValidator(
+              new PipeParameters(
+                  new HashMap<String, String>() {
+                    {
+                      put(PipeExtractorConstant.EXTRACTOR_PATTERN_KEY, "root.a-b");
+                    }
+                  })));
+    } catch (Exception e) {
+      Assert.assertEquals(e.getClass(), IllegalPathException.class);
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/test/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractorTest.java:71]: @Test(expected = IllegalArgumentException.class)

**Review by Caideyipi — COMMENTED** [review]
PTAL

File: iotdb-core/datanode/src/test/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractorTest.java [hunk 3]
```diff
@@ -53,4 +53,32 @@ public void testIoTDBDataRegionExtractor() {
...
+  @Test
+  public void testIoTDBDataRegionExtractorWithPattern() {
+    Assert.assertEquals(
+        testIoTDBDataRegionExtractorWithPattern("root.a-b").getClass(),
+        IllegalArgumentException.class);
+    Assert.assertEquals(
+        testIoTDBDataRegionExtractorWithPattern("root.1.a").getClass(),
+        IllegalArgumentException.class);
+    Assert.assertNull(testIoTDBDataRegionExtractorWithPattern("root.`a-b`"));
+    Assert.assertNull(testIoTDBDataRegionExtractorWithPattern("root.1"));
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/test/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractorTest.java:66]:
```suggestion
    Assert.assertEquals(
        testIoTDBDataRegionExtractorWithPattern("root.a-b").getClass(),
        IllegalArgumentException.class);
    Assert.assertEquals(
        testIoTDBDataRegionExtractorWithPattern("root.1.a").getClass(),
        IllegalArgumentException.class);
    Assert.assertNull(testIoTDBDataRegionExtractorWithPattern("root.`a-b`"));
    Assert.assertNull(testIoTDBDataRegionExtractorWithPattern("root.1"));
    // root
    // r
    // ""
    // 123
    // root.ab
    // root.a.b.c.1e2
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java [hunk 4]
```diff
@@ -84,6 +90,29 @@ public IoTDBDataRegionExtractor() {
...
+        validator
+            .getParameters()
+            .getStringOrDefault(
+                Arrays.asList(EXTRACTOR_PATTERN_KEY, SOURCE_PATTERN_KEY),
+                EXTRACTOR_PATTERN_DEFAULT_VALUE);
+    try {
+      PathUtils.isLegalPath(pattern);
+    } catch (IllegalPathException e) {
+      try {
+        String[] pathNodes = StringUtils.split(pattern, "\\.");
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java:104]: What if pathNodes.length == 0 or == 1

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java [hunk 5]
```diff
@@ -84,6 +90,29 @@ public IoTDBDataRegionExtractor() {
 
   @Override
   public void validate(PipeParameterValidator validator) throws Exception {
+    // Check whether the pattern is legal
+    String pattern =
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java:94]:
```suggestion
    final String pattern =
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java [hunk 6]
```diff
@@ -84,6 +90,29 @@ public IoTDBDataRegionExtractor() {
...
+    } catch (IllegalPathException e) {
+      try {
+        String[] pathNodes = StringUtils.split(pattern, "\\.");
+        PathUtils.splitPathToDetachedNodes(
+            String.join(".", Arrays.copyOfRange(pathNodes, 0, pathNodes.length - 1)));
+        String lastNode = pathNodes[pathNodes.length - 1];
+        if (!"".equals(lastNode)) {
+          Double.parseDouble(lastNode);
+        }
+      } catch (NumberFormatException | IllegalPathException ignored) {
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java:111]:
```suggestion
      } catch (Exception e) {
        LOGGER.warn("");
        throw new IllegalArgumentException(pattern);
      }
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java [hunk 7]
```diff
@@ -84,6 +90,29 @@ public IoTDBDataRegionExtractor() {
 
   @Override
   public void validate(PipeParameterValidator validator) throws Exception {
+    // Check whether the pattern is legal
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/IoTDBDataRegionExtractor.java:93]: make it a seperate function.
 
validatePattern(validator
  .getParameters()
  .getStringOrDefault(
     Arrays.asList(EXTRACTOR_PATTERN_KEY, SOURCE_PATTERN_KEY),
     EXTRACTOR_PATTERN_DEFAULT_VALUE));

**Review by SteveYurongSu — APPROVED** [review]

**Pull request merged**